File: algorithms/sorting/bubble-sort/bubble_sort.py
"""Reference Bubble Sort implementation used by Algorithm Lab."""


def bubble_sort(numbers):
    """Return a sorted copy of numbers without mutating the input."""
    array = numbers.copy()

    for last in range(len(array) - 1, 0, -1):
        swapped = False

        for index in range(last):
            if array[index] > array[index + 1]:
                array[index], array[index + 1] = array[index + 1], array[index]
                swapped = True

        if not swapped:
            break

    return array

File: algorithms/sorting/bubble-sort/BubbleSort.java
import java.util.Arrays;

/** Reference Bubble Sort implementation used by Algorithm Lab. */
public final class BubbleSort {
    private BubbleSort() {
    }

    /** Return a sorted copy of the input without mutating it. */
    public static int[] bubbleSort(int[] numbers) {
        int[] array = Arrays.copyOf(numbers, numbers.length);

        for (int pass = 0; pass < array.length - 1; pass++) {
            boolean swapped = false;

            for (int index = 0; index < array.length - pass - 1; index++) {
                if (array[index] > array[index + 1]) {
                    int temporary = array[index];
                    array[index] = array[index + 1];
                    array[index + 1] = temporary;
                    swapped = true;
                }
            }

            if (!swapped) {
                break;
            }
        }

        return array;
    }
}
